File: examples/llm-proxies/openai-gpt4-proxy.js
#!/usr/bin/env node

/**
 * OpenAI GPT-4 Proxy for Gemini CLI
 * 
 * This proxy intercepts Gemini API calls and redirects them to OpenAI's GPT-4 API.
 * It translates between Gemini's API format and OpenAI's format.
 * 
 * Usage: node openai-gpt4-proxy.js --port 8080
 * 
 * Environment Variables:
 * - OPENAI_API_KEY: Your OpenAI API key
 * - OPENAI_MODEL: Model to use (default: gpt-4o)
 * - OPENAI_BASE_URL: Custom OpenAI-compatible endpoint (optional)
 */

const http = require('http');
const https = require('https');
const url = require('url');
const crypto = require('crypto');

// Configuration
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
const OPENAI_MODEL = process.env.OPENAI_MODEL || 'gpt-4o';
const OPENAI_BASE_URL = process.env.OPENAI_BASE_URL || 'https://api.openai.com';

if (!OPENAI_API_KEY) {
    console.error('❌ ERROR: OPENAI_API_KEY environment variable is required');
    process.exit(1);
}

// Parse command line arguments
const args = process.argv.slice(2);
const portIndex = args.indexOf('--port');
const PORT = portIndex !== -1 ? parseInt(args[portIndex + 1]) : 8080;

console.log(`🟨 OpenAI GPT-4 Proxy starting on port ${PORT}`);
console.log(`📡 Target: ${OPENAI_BASE_URL} (${OPENAI_MODEL})`);

/**
 * Convert Gemini request format to OpenAI format
 */
function convertGeminiToOpenAI(geminiRequest) {
    try {
        const { contents, tools, generationConfig } = geminiRequest;
        
        const messages = [];
        
        // Convert Gemini contents to OpenAI messages
        if (contents && Array.isArray(contents)) {
            for (const content of contents) {
                if (content.role === 'user' || content.role === 'model') {
                    const role = content.role === 'model' ? 'assistant' : content.role;
                    
                    if (content.parts && Array.isArray(content.parts)) {
                        let messageContent = '';
                        for (const part of content.parts) {
                            if (part.text) {
                                messageContent += part.text;
                            }
                            // Handle function calls and other part types
                            if (part.functionCall) {
                                messageContent += `\n[Function Call: ${part.functionCall.name}]`;
                            }
                        }
                        
                        if (messageContent.trim()) {
                            messages.push({
                                role: role,
                                content: messageContent.trim()
                            });
                        }
                    }
                }
            }
        }
        
        const openaiRequest = {
            model: OPENAI_MODEL,
            messages: messages,
            stream: false,
            temperature: generationConfig?.temperature || 0.1,
            max_tokens: generationConfig?.maxOutputTokens || 4000,
        };
        
        // Add tools if present
        if (tools && Array.isArray(tools) && tools.length > 0) {
            const openaiTools = tools.map(tool => ({
                type: "function",
                function: {
                    name: tool.functionDeclarations?.[0]?.name || 'unknown',
                    description: tool.functionDeclarations?.[0]?.description || '',
                    parameters: tool.functionDeclarations?.[0]?.parameters || {}
                }
            }));
            
            if (openaiTools.length > 0) {
                openaiRequest.tools = openaiTools;
                openaiRequest.tool_choice = "auto";
            }
        }
        
        return openaiRequest;
    } catch (error) {
        console.error('❌ Error converting Gemini to OpenAI format:', error);
        throw error;
    }
}

/**
 * Convert OpenAI response format to Gemini format
 */
function convertOpenAIToGemini(openaiResponse) {
    try {
        const choice = openaiResponse.choices?.[0];
        if (!choice) {
            throw new Error('No choices in OpenAI response');
        }
        
        const geminiResponse = {
            candidates: [{
                content: {
                    parts: [{
                        text: choice.message?.content || ''
                    }],
                    role: 'model'
                },
                finishReason: choice.finish_reason === 'stop' ? 'STOP' : 'OTHER',
                index: 0,
                safetyRatings: []
            }],
            usageMetadata: {
                promptTokenCount: openaiResponse.usage?.prompt_tokens || 0,
                candidatesTokenCount: openaiResponse.usage?.completion_tokens || 0,
                totalTokenCount: openaiResponse.usage?.total_tokens || 0
            }
        };
        
        // Handle function calls
        if (choice.message?.tool_calls) {
            geminiResponse.candidates[0].content.parts = choice.message.tool_calls.map(call => ({
                functionCall: {
                    name: call.function.name,
                    args: JSON.parse(call.function.arguments || '{}')
                }
            }));
        }
        
        return geminiResponse;
    } catch (error) {
        console.error('❌ Error converting OpenAI to Gemini format:', error);
        throw error;
    }
}

/**
 * Make request to OpenAI API
 */
function makeOpenAIRequest(openaiRequest) {
    return new Promise((resolve, reject) => {
        const requestData = JSON.stringify(openaiRequest);
        const parsedUrl = new URL(`${OPENAI_BASE_URL}/v1/chat/completions`);
        
        const options = {
            hostname: parsedUrl.hostname,
            port: parsedUrl.port || (parsedUrl.protocol === 'https:' ? 443 : 80),
            path: parsedUrl.pathname,
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${OPENAI_API_KEY}`,
                'Content-Length': Buffer.byteLength(requestData),
                'User-Agent': 'DMTools-LLM-Proxy/1.0'
            }
        };
        
        const client = parsedUrl.protocol === 'https:' ? https : http;
        const req = client.request(options, (res) => {
            let data = '';
            
            res.on('data', (chunk) => {
                data += chunk;
            });
            
            res.on('end', () => {
                try {
                    const response = JSON.parse(data);
                    if (res.statusCode >= 200 && res.statusCode < 300) {
                        resolve(response);
                    } else {
                        reject(new Error(`OpenAI API error: ${res.statusCode} - ${response.error?.message || data}`));
                    }
                } catch (error) {
                    reject(new Error(`Failed to parse OpenAI response: ${error.message}`));
                }
            });
        });
        
        req.on('error', (error) => {
            reject(new Error(`OpenAI request failed: ${error.message}`));
        });
        
        req.write(requestData);
        req.end();
    });
}

/**
 * HTTP Proxy Server
 */
const server = http.createServer(async (req, res) => {
    // Enable CORS
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
    
    // Health check endpoint
    if (req.method === 'GET' && req.url === '/health') {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ 
            status: 'healthy', 
            proxy: 'openai-gpt4',
            model: OPENAI_MODEL,
            timestamp: new Date().toISOString()
        }));
        return;
    }
    
    // Handle OPTIONS requests
    if (req.method === 'OPTIONS') {
        res.writeHead(200);
        res.end();
        return;
    }
    
    // Only handle POST requests to Gemini API endpoints
    if (req.method !== 'POST') {
        res.writeHead(405, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: 'Method not allowed' }));
        return;
    }
    
    // Check if this is a Gemini API request
    const urlPath = url.parse(req.url).pathname;
    if (!urlPath.includes('generateContent') && !urlPath.includes('/v1/models/') && !urlPath.includes('/v1beta/')) {
        res.writeHead(404, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: 'Not a Gemini API endpoint' }));
        return;
    }
    
    // Collect request body
    let body = '';
    req.on('data', chunk => {
        body += chunk.toString();
    });
    
    req.on('end', async () => {
        try {
            console.log(`🔄 Intercepting Gemini API call: ${urlPath}`);
            
            const geminiRequest = JSON.parse(body);
            const openaiRequest = convertGeminiToOpenAI(geminiRequest);
            
            console.log(`📤 Forwarding to OpenAI (${OPENAI_MODEL})`);
            const openaiResponse = await makeOpenAIRequest(openaiRequest);
            
            const geminiResponse = convertOpenAIToGemini(openaiResponse);
            
            console.log(`✅ Request completed successfully`);
            
            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify(geminiResponse));
            
        } catch (error) {
            console.error('❌ Proxy error:', error.message);
            
            res.writeHead(500, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({
                error: {
                    code: 500,
                    message: `Proxy error: ${error.message}`,
                    status: 'INTERNAL_ERROR'
                }
            }));
        }
    });
});

// Start server
server.listen(PORT, () => {
    console.log(`✅ OpenAI GPT-4 Proxy running on http://localhost:${PORT}`);
    console.log(`💡 Health check: http://localhost:${PORT}/health`);
});

// Graceful shutdown
process.on('SIGINT', () => {
    console.log('\n🛑 Shutting down proxy server...');
    server.close(() => {
        console.log('✅ Proxy server stopped');
        process.exit(0);
    });
});

process.on('SIGTERM', () => {
    console.log('\n🛑 Shutting down proxy server...');
    server.close(() => {
        console.log('✅ Proxy server stopped');
        process.exit(0);
    });
});

File: examples/llm-proxies/openai-gpt4-proxy.py
#!/usr/bin/env python3

"""
OpenAI GPT-4 Proxy for Gemini CLI

This proxy intercepts Gemini API calls and redirects them to OpenAI's GPT-4 API.
It translates between Gemini's API format and OpenAI's format.

Usage: python3 openai-gpt4-proxy.py --port 8080

Environment Variables:
- OPENAI_API_KEY: Your OpenAI API key
- OPENAI_MODEL: Model to use (default: gpt-4o)
- OPENAI_BASE_URL: Custom OpenAI-compatible endpoint (optional)
"""

import argparse
import os
import json
import signal
import sys
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import requests
from datetime import datetime
import threading
import time

# Configuration
OPENAI_API_KEY = os.getenv('OPENAI_API_KEY')
OPENAI_MODEL = os.getenv('OPENAI_MODEL', 'gpt-4o')
OPENAI_BASE_URL = os.getenv('OPENAI_BASE_URL', 'https://api.openai.com')

if not OPENAI_API_KEY:
    print('❌ ERROR: OPENAI_API_KEY environment variable is required')
    sys.exit(1)

class LLMProxyHandler(BaseHTTPRequestHandler):
    """HTTP request handler for the LLM proxy"""
    
    def log_message(self, format, *args):
        """Override to control logging"""
        pass  # Disable default HTTP logging
    
    def do_OPTIONS(self):
        """Handle CORS preflight requests"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
        self.end_headers()
    
    def do_GET(self):
        """Handle GET requests - mainly health checks"""
        if self.path == '/health':
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            
            health_response = {
                'status': 'healthy',
                'proxy': 'openai-gpt4',
                'model': OPENAI_MODEL,
                'timestamp': datetime.now().isoformat()
            }
            self.wfile.write(json.dumps(health_response).encode())
        else:
            self.send_response(404)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Not found'}).encode())
    
    def do_POST(self):
        """Handle POST requests - proxy to OpenAI"""
        try:
            # Check if this is a Gemini API request
            if not self._is_gemini_api_request():
                self.send_response(404)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Not a Gemini API endpoint'}).encode())
                return
            
            # Read request body
            content_length = int(self.headers.get('Content-Length', 0))
            request_body = self.rfile.read(content_length).decode('utf-8')
            
            print(f'🔄 Intercepting Gemini API call: {self.path}')
            
            # Parse and convert request
            gemini_request = json.loads(request_body)
            openai_request = self._convert_gemini_to_openai(gemini_request)
            
            print(f'📤 Forwarding to OpenAI ({OPENAI_MODEL})')
            
            # Make request to OpenAI
            openai_response = self._make_openai_request(openai_request)
            
            # Convert response back to Gemini format
            gemini_response = self._convert_openai_to_gemini(openai_response)
            
            print('✅ Request completed successfully')
            
            # Send response
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(gemini_response).encode())
            
        except Exception as e:
            print(f'❌ Proxy error: {str(e)}')
            
            error_response = {
                'error': {
                    'code': 500,
                    'message': f'Proxy error: {str(e)}',
                    'status': 'INTERNAL_ERROR'
                }
            }
            
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(error_response).encode())
    
    def _is_gemini_api_request(self):
        """Check if the request is for a Gemini API endpoint"""
        return ('generateContent' in self.path or 
                '/v1/models/' in self.path or 
                '/v1beta/' in self.path)
    
    def _convert_gemini_to_openai(self, gemini_request):
        """Convert Gemini request format to OpenAI format"""
        try:
            contents = gemini_request.get('contents', [])
            tools = gemini_request.get('tools', [])
            generation_config = gemini_request.get('generationConfig', {})
            
            messages = []
            
            # Convert Gemini contents to OpenAI messages
            for content in contents:
                role = content.get('role')
                if role in ['user', 'model']:
                    openai_role = 'assistant' if role == 'model' else role
                    
                    parts = content.get('parts', [])
                    message_content = ''
                    
                    for part in parts:
                        if 'text' in part:
                            message_content += part['text']
                        if 'functionCall' in part:
                            message_content += f"\n[Function Call: {part['functionCall'].get('name', 'unknown')}]"
                    
                    if message_content.strip():
                        messages.append({
                            'role': openai_role,
                            'content': message_content.strip()
                        })
            
            openai_request = {
                'model': OPENAI_MODEL,
                'messages': messages,
                'stream': False,
                'temperature': generation_config.get('temperature', 0.1),
                'max_tokens': generation_config.get('maxOutputTokens', 4000),
            }
            
            # Add tools if present
            if tools:
                openai_tools = []
                for tool in tools:
                    function_declarations = tool.get('functionDeclarations', [])
                    for func_decl in function_declarations:
                        openai_tools.append({
                            'type': 'function',
                            'function': {
                                'name': func_decl.get('name', 'unknown'),
                                'description': func_decl.get('description', ''),
                                'parameters': func_decl.get('parameters', {})
                            }
                        })
                
                if openai_tools:
                    openai_request['tools'] = openai_tools
                    openai_request['tool_choice'] = 'auto'
            
            return openai_request
            
        except Exception as e:
            raise Exception(f'Error converting Gemini to OpenAI format: {str(e)}')
    
    def _convert_openai_to_gemini(self, openai_response):
        """Convert OpenAI response format to Gemini format"""
        try:
            choices = openai_response.get('choices', [])
            if not choices:
                raise Exception('No choices in OpenAI response')
            
            choice = choices[0]
            message = choice.get('message', {})
            
            parts = []
            
            # Handle regular text response
            if message.get('content'):
                parts.append({'text': message['content']})
            
            # Handle function calls
            if message.get('tool_calls'):
                for tool_call in message['tool_calls']:
                    function = tool_call.get('function', {})
                    try:
                        args = json.loads(function.get('arguments', '{}'))
                    except:
                        args = {}
                    
                    parts.append({
                        'functionCall': {
                            'name': function.get('name', 'unknown'),
                            'args': args
                        }
                    })
            
            finish_reason = choice.get('finish_reason', 'stop')
            gemini_finish_reason = 'STOP' if finish_reason == 'stop' else 'OTHER'
            
            usage = openai_response.get('usage', {})
            
            gemini_response = {
                'candidates': [{
                    'content': {
                        'parts': parts,
                        'role': 'model'
                    },
                    'finishReason': gemini_finish_reason,
                    'index': 0,
                    'safetyRatings': []
                }],
                'usageMetadata': {
                    'promptTokenCount': usage.get('prompt_tokens', 0),
                    'candidatesTokenCount': usage.get('completion_tokens', 0),
                    'totalTokenCount': usage.get('total_tokens', 0)
                }
            }
            
            return gemini_response
            
        except Exception as e:
            raise Exception(f'Error converting OpenAI to Gemini format: {str(e)}')
    
    def _make_openai_request(self, openai_request):
        """Make request to OpenAI API"""
        try:
            headers = {
                'Content-Type': 'application/json',
                'Authorization': f'Bearer {OPENAI_API_KEY}',
                'User-Agent': 'DMTools-LLM-Proxy/1.0'
            }
            
            url = f'{OPENAI_BASE_URL}/v1/chat/completions'
            
            response = requests.post(
                url,
                headers=headers,
                json=openai_request,
                timeout=60
            )
            
            if response.status_code != 200:
                error_data = response.json() if response.headers.get('content-type', '').startswith('application/json') else response.text
                raise Exception(f'OpenAI API error: {response.status_code} - {error_data}')
            
            return response.json()
            
        except requests.exceptions.RequestException as e:
            raise Exception(f'OpenAI request failed: {str(e)}')

def signal_handler(signum, frame):
    """Handle shutdown signals"""
    print('\n🛑 Shutting down proxy server...')
    sys.exit(0)

def main():
    parser = argparse.ArgumentParser(description='OpenAI GPT-4 Proxy for Gemini CLI')
    parser.add_argument('--port', type=int, default=8080, help='Port to run the proxy server on')
    args = parser.parse_args()
    
    print(f'🐍 OpenAI GPT-4 Proxy starting on port {args.port}')
    print(f'📡 Target: {OPENAI_BASE_URL} ({OPENAI_MODEL})')
    
    # Set up signal handlers for graceful shutdown
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    # Create and start server
    server = HTTPServer(('localhost', args.port), LLMProxyHandler)
    
    try:
        print(f'✅ OpenAI GPT-4 Proxy running on http://localhost:{args.port}')
        print(f'💡 Health check: http://localhost:{args.port}/health')
        server.serve_forever()
    except KeyboardInterrupt:
        print('\n🛑 Shutting down proxy server...')
    finally:
        server.server_close()
        print('✅ Proxy server stopped')

if __name__ == '__main__':
    main()
